infra/Pos/DHT/Real/Real.hs

```diff
@@ -138,40 +140,40 @@ getKnownPeersImpl = do
     (inst, initialPeers, explicitInitial) <-
         KademliaDHT $
         (,,) <$> ask <*> asks kdiInitialPeers <*> asks kdiExplicitInitial
-    peers <- liftIO $ K.dumpPeers $ kdiHandle inst
+    buckets <- liftIO (K.viewBuckets $ kdiHandle inst)
     let initPeers = bool [] initialPeers explicitInitial
-    filter ((/= myId) . dhtNodeId) <$> extendPeers inst myId initPeers peers
+    filter ((/= myId) . dhtNodeId) <$> extendPeers inst myId initPeers buckets
   where
-    extendPeers inst myId initial peers =
+    extendPeers
+        :: MonadIO m1
+        => KademliaDHTInstance
+        -> DHTKey
+        -> [DHTNode]
+        -> [[(K.Node DHTKey, Int64)]]
+        -> m1 [DHTNode]
+    extendPeers inst myId initial buckets =
         map snd .
         HM.toList .
         HM.delete myId .
         flip (foldr $ \n -> HM.insert (dhtNodeId n) n) initial .
         HM.fromList . map (\(toDHTNode -> n) -> (dhtNodeId n, n)) <$>
-        (updateCache inst =<< selectSufficientNodes inst myId peers)
-    selectSufficientNodes inst myId l =
```

Comment:
Simple things have been written so complicated way.
There are buckets in Kademlia, previous code took them and concatenate. And much logic was be related to split concatenated buckets on buckets again (but using other custom logic).